File: server/app/modules/wallet/schemas/wallet_schema.py
from datetime import datetime
from uuid import UUID

from pydantic import field_validator
from sqlmodel import SQLModel


class OwnerWalletBalanceRead(SQLModel):
    wallet_id: UUID
    repair_shop_id: UUID
    balance: float
    updated_at: datetime


class OwnerWalletTransactionRead(SQLModel):
    id: UUID
    type: str
    status: str
    amount: float
    balance_before: float
    balance_after: float
    description: str | None = None
    created_at: datetime


class OwnerWalletTransactionsResponse(SQLModel):
    transactions: list[OwnerWalletTransactionRead]


class OwnerWalletTopupQrCreateRequest(SQLModel):
    amount: float

    @field_validator("amount")
    @classmethod
    def validate_amount(cls, value: float) -> float:
        if value <= 0:
            raise ValueError("El monto debe ser mayor a 0")

        cents = round(value * 100)
        if abs(value * 100 - cents) > 1e-6:
            raise ValueError("El monto debe tener maximo 2 decimales")

        return value


class OwnerWalletTopupQrRead(SQLModel):
    transaction_id: UUID
    internal_reference: str
    amount: float
    wallet_id: UUID
    repair_shop_id: UUID
    generated_at: datetime
    expires_at: datetime
    status: str
    qr_payload: str
    qr_image_url: str


class OwnerWalletTopupConfirmResponse(SQLModel):
    transaction_id: UUID
    status: str
    previous_balance: float
    new_balance: float

File: web/src/app/features/wallet/data/schemas/wallet.schema.ts
import { z } from 'zod';

export const OwnerWalletBalanceSchema = z.object({
  wallet_id: z.string().uuid(),
  repair_shop_id: z.string().uuid(),
  balance: z.number(),
  updated_at: z.string(),
});

export const OwnerWalletTransactionSchema = z.object({
  id: z.string().uuid(),
  type: z.string(),
  status: z.string(),
  amount: z.number(),
  balance_before: z.number(),
  balance_after: z.number(),
  description: z.string().nullable().optional(),
  created_at: z.string(),
});

export const OwnerWalletTransactionsResponseSchema = z.object({
  transactions: z.array(OwnerWalletTransactionSchema),
});

export const OwnerWalletTopupQrRequestSchema = z.object({
  amount: z.number().positive('El monto debe ser mayor a 0'),
});

export const OwnerWalletTopupQrSchema = z.object({
  transaction_id: z.string().uuid(),
  internal_reference: z.string(),
  amount: z.number(),
  wallet_id: z.string().uuid(),
  repair_shop_id: z.string().uuid(),
  generated_at: z.string(),
  expires_at: z.string(),
  status: z.string(),
  qr_payload: z.string(),
  qr_image_url: z.string().url(),
});

export const OwnerWalletTopupConfirmSchema = z.object({
  transaction_id: z.string().uuid(),
  status: z.string(),
  previous_balance: z.number(),
  new_balance: z.number(),
});

export type OwnerWalletBalanceData = z.infer<typeof OwnerWalletBalanceSchema>;
export type OwnerWalletTransactionData = z.infer<typeof OwnerWalletTransactionSchema>;
export type OwnerWalletTopupQrRequestData = z.infer<typeof OwnerWalletTopupQrRequestSchema>;
export type OwnerWalletTopupQrData = z.infer<typeof OwnerWalletTopupQrSchema>;
export type OwnerWalletTopupConfirmData = z.infer<typeof OwnerWalletTopupConfirmSchema>;
